beacon_chain/spec/defects.nim

```
@@ -0,0 +1,92 @@
+  RecordDefect* = object of ExitProcessDefect
+  Web3CmdDefect* = object of ExitProcessDefect
+  SlashInterDefect* = object of ExitProcessDefect
+
+  ValidatorClientDefect* = object of ExitProcessDefect
+  LightClientDefect* = object of ExitProcessDefect
+  DepositContractDefect* = object of ExitProcessDefect
+
+template declareExitHelpers(defect, code: untyped): untyped =
+  template `raise defect`*() =
+    raise (ref defect)(exitCode: code, msg: "")
+  template `raise defect`*(message: string) =
+    raise (ref defect)(exitCode: code, msg: message)
+
+# Submodules exits
+declareExitHelpers(SuccessDefect, 0)
+declareExitHelpers(DatabaseDefect, 12)
+declareExitHelpers(SlashingDefect, 13)
+declareExitHelpers(NetworkDefect, 14)
+declareExitHelpers(ClearanceDefect, 15)
+declareExitHelpers(StrictDefect, 16)
+declareExitHelpers(DagDefect, 17)
+declareExitHelpers(KeystoreDefect, 18)
+declareExitHelpers(ProcessorDefect, 19)
+declareExitHelpers(DiscoveryDefect, 20)
+declareExitHelpers(ChainListDefect, 21)
+declareExitHelpers(OverseerDefect, 22)
+declareExitHelpers(LogDefect, 23)
+declareExitHelpers(ConfigDefect, 24)
+declareExitHelpers(KeymanagerDefect, 25)
+declareExitHelpers(MetadataDefect, 26)
+# Modules exits
+declareExitHelpers(DepositsDefect, 30)
+declareExitHelpers(TrustedSyncDefect, 31)
+declareExitHelpers(WalletsDefect, 32)
+declareExitHelpers(BeaconNodeDefect, 33)
+declareExitHelpers(RecordDefect, 34)
+declareExitHelpers(Web3CmdDefect, 35)
+declareExitHelpers(SlashInterDefect, 36)
+declareExitHelpers(ValidatorClientDefect, 37)
+declareExitHelpers(LightClientDefect, 38)
+declareExitHelpers(DepositContractDefect, 39)
+
+template withDefectsHandlers*(pbody, fbody: untyped) =
+  try:
+    pbody
+  except ExitProcessDefect as exc:
+    # This defects are already logged, so we just quit.
+    quit(exc.exitCode)
+  finally:
+    fbody
+
+template withDefectsHandlers*(pbody: untyped) =
+  try:
+    pbody
+  except ExitProcessDefect as exc:
```

Catching `Defect` is [undefined behavior](https://nim-lang.org/docs/manual.html#definitions) - this is not something we want to use/rely on broadly - it can change meaning in future nim versions and/or not behave properly . 

Rather than defect, in order to avoid `quit`, the fix would be to return errors from the functions and deal with them _or_ let the defect be raised and not caught.

---

Its first time i see that catching `Defect` is `undefined behavior`, because in such case generating `Defect` is also `undefined behavior` and code which was written with `Result.expect()` and `quit X` in mind should be totally rewritten before attempts to combine CL and EL into single binary. Otherwise `undefined behavior` created by CL could cause significant losses in EL (e.g. database could be left in `undefined state`).

---

> Otherwise `undefined behavior` created by CL could cause significant losses in EL (e.g. database could be left in `undefined state`).

ELs should not be able to put into undefined states by anything the engine API does. That's an EL bug.

---

> generating Defect is also undefined behavior

a raised `Defect` is guaranteed to terminate the application - how it does that is undefined - ie `except Defect` may or may not work, ie you can't expect that `quit(exc.exitCode)` or anything inside this exception handler will actually run. 

This undefined behavior is why we don't use `except:` and `except Exception:` - they are both undefined as a consequence of catching `Defect` being undefined (I consider it a design flaw in the language).

The third correct option is to turn `ExitProcessDefect` into a `CatchableError`. 

From a code perspective, it's also bad to have untracked exceptions / hidden control flows like this - this goes back to our original decision to use `raises` consistently - `raises` tracking does not work with `Defect` either.

`excpect` is the same as `doAssert`, but with a better error message - it works as intended, ie it terminates the application.

---

> (e.g. database could be left in undefined state)

generally, all our database interactions happen in a transaction - the worst thing that can happen is a rollback.